File: 프로그래머스/0/120861. 캐릭터의 좌표/캐릭터의 좌표.js
function solution(keyinput, board) {
    let [x, y] = [0, 0]
    
    const offSet = [(board[0] - 1) / 2, (board[1] - 1) / 2];
    
    for(const d of keyinput) {
        if(d === 'left') {
            if(x - 1 >= -offSet[0]) x--;
        }
        if(d === 'right') {
            if(x + 1 <= offSet[0]) x++;
        }
        if(d === 'up') {
            if(y + 1 <= offSet[1]) y++;
        }
        if(d === 'down') {
            if(y - 1 >= -offSet[1]) y--;
        }
    }
    
    return [x, y];
}

File: 프로그래머스/0/120861. 캐릭터의 좌표/캐릭터의 좌표.py
def move(order, x, y):
    if order == 'left':
        return x - 1, y
    if order == 'right':
        return x + 1, y
    if order == 'up':
        return x, y + 1
    if order == 'down':
        return x, y - 1

def solution(keyinput, board):
    x_left, x_right = -(board[0] - 1) // 2, (board[0] - 1) // 2
    y_down, y_up = -(board[1] - 1) // 2, (board[1] - 1) // 2
    def is_valid_move(x, y):
        return x_left <= x <= x_right and y_down <= y <= y_up
    
    X, Y = 0, 0
    for order in keyinput:
        nx, ny = move(order, X, Y)
        if is_valid_move(nx, ny):
            X = nx
            Y = ny
    return [X, Y]